### ISSUE TITLE
Stop time of last component in routine is incorrect

### ISSUE THREAD

**Issue by pn2200** [psychopy/psychopy#3672]
I have an experiment displaying a ShapeStim for one second, then a TextStim for 4 seconds, in a loop five times. The output for the start and stop times of the TextStim is incorrect: the `.stopped` field has value `None` for some of the entries (sometimes 2 entries, sometimes 3), and the remainder are values before the `.started` field. When `TextStim.stopped` does have a value, then it is equal to `ShapeStim.started`.

The TextStim's `.started` field is before the end of the ShapeStim's `.stopped` field; it looks like the TextStim starts one or two frames before the ShapeStim ends, so they overlap visually when the experiment is run. In this case, it might be that the time is accurately recorded, but it does not reflect the desired experiment flow where the ShapeStim is displayed, followed by the TextStim.

I have attached the experiment file, the conditions file, and a sample output file. They were created using PsychoPy 2021.1.2.

[overlap.zip](https://github.com/psychopy/psychopy/files/6104399/overlap.zip)

**Comment by pn2200** [comment]
I think this issue affects more components than just `TextStim`. I think that the stop times of the last component in any routine is not recorded correctly.

In this specific experiment, this issue occurs because the `routineTimer` becomes negative (`routineTimer.getTime()` returns a value less than 0) while the TextStim has a status of STARTED and `tThisFlipGlobal` is less than the TextStim's start time plus duration (`tThisFlipGlobal > text.tStartRefresh + 4.0-frameTolerance` is false), so code that stores the stop time for the component is never executed.

I don't know the right way to resolve this issue. Adding an else-statement on the while loop like the following seems to work, so I will create a pull request based on this:  

```
while continueRoutine and routineTimer.getTime() > 0:
    # Run loop
else:
    for thisComponent in trialComponents:
        if hasattr(thisComponent, 'status') and thisComponent.status == STARTED:
            # Stop component and record stopping time
            thisComponent.tStop = t
            thisComponent.frameNStop = frameN
            win.timeOnFlip(thisComponent, 'tStopRefresh')
    win.flip()
```

**Comment by pn2200** [comment]
I also think that https://github.com/psychopy/psychopy/issues/2916 is a special case of this more general problem, so I think that the fix for that issue should be revisited.

**Comment by peircej** [comment]
So, the general issue about the stop times for last thing in a routine is that the next frame flip typically occurs after we've gone around to the next entry of the trials loop. Unfortunately, in TrialHandler each trial is not its own object (unlike in TrialHandler2), rather it's an index into a matrix. That means we don't have the option to add data to the *previous* trial easily once it's passed. So by the time the true end time is know the ability to save it has been lost. Your solution of adding a win.flip is one way to go but not something we'd want to do as a general solution (because it means you always get a blank screen at the end of the routine, which might not be wanted)

**Comment by peircej** [comment]
Basically, I think the true solution is that we switch to using TrialHandler2 which manages each trial essentially as its own dict that could then be modified later. That switch to the new class is held up by not having quite a full set of comparable methods in TrialHandler2 (it can't save the summary xlsx output that some users prefer).

I should note though, that the easy workaround is that the onset time of a stimulus at t=0 in the next routine will be the true offset time of the stimulus in the previous, if that makes sense. So the data are always there

**Comment by pn2200** [comment]
> I should note though, that the easy workaround is that the onset time of a stimulus at t=0 in the next routine will be the true offset time of the stimulus in the previous, if that makes sense. So the data are always there

This assumes there is a stimulus at t=0 in the next routine. That isn't required (I could choose to leave the screen blank for a bit and start the stimulus after t=0). Also, the stop time of the last routine of the last trial is completely lost. There is no next routine and there is no next trial, so there is no way to recover that data.

So it looks like the way forward is to fix #1590 first, then replace the use of `TrialHandler` with `TrialHandler2`. I will get to work on it. Are there any other issues with`TrialHandler2` that need to be addressed before it can be used as a replacement?

**Comment by peircej** [comment]
> This assumes there is a stimulus at t=0 in the next routine. 

Yes, indeed, but one could add something for timing measures if needed. What I mean really is that this problem has a potential workaround whereas adding win.flip() will be wrong for some people and they *won't* have a workaround (which seems worse)

> So it looks like the way forward is to fix #1590 first

Yes, I think we should try and fix TrialHandler2 and use that instead. Did you fancy attacking that Issue?  I'd be delighted to help If you had time to nail it down. I mean, if implementing that method is too hard (it might be a real pain) then we could also:
1. essentially deprecate that summarised file output in the Builder view so that it couldn't be selected in new experiments
2. test if it's there anyway at script-write time and use TrialHandler2 *unless* the user has requested excel summarised output

Then we would gradually move people away from this output (it feels outdated anyway - it's a bit easy for excel users but worse for someone using R, say, because 1 row per condition rather than 1 row per trial) but people with existing experiments would not have their code broken.

Actually, as I write this it feels like the right way to go! 😄

**Comment by pn2200** [comment]
I have been looking at implementing `saveAsExcel` for `TrialHandler2`. It seem possible to do, and then there could be a replacement of `TrialHandler` with `TrialHandler2` when Builder generates scripts, and TrialHandler could be marked as deprecated.

However, just using `TrialHandler2` in this experiment that does not record stop times of the last component, does not resolve this issue, because the component stop time is stored only on the next flip, which does not happen in the current routing. I am getting hung up on that part, and could use some assistance.

**Comment by peircej** [comment]
This was fixed in 2022.2 release which doesn't actually use TrialHandler2 but uses the ExperimentHandler currentTrial property to fetch a stored version of the current trial that can still be updated after the trial ends
